File: app/services/stot_engine.py
"""
STOTEngine — CLP (AT) No 1272/2008 Ek I §3.9
STOT RE (Tekrarlı Maruziyet) karışım hesaplama

H372 (STOT RE 1): organ Cat1 toplamı ≥ %10.0 (generic) VEYA madde SCL_H372 ≤ konsantrasyon
H373 (STOT RE 2): Cat1 %1.0–%10.0 VEYA Cat2 ≥ %10.0 (generic) VEYA SCL_H373 ≤ konsantrasyon
SCL generic eşiğin önüne geçer — CLP Madde 10(3) / KKDİK Ek-2 Madde 10
Kaynak: CLP Ek-1 §3.9, Tablo 3.9.4
"""

from typing import List, Dict

# ─── SABİTLER (stot_re_service'den taşındı) ──────────────────────────────────

GENERAL_ORGAN = 'Genel (organ belirsiz)'  # Organ belirtilmemiş sentinel

ORGAN_ALIASES = {
    'blood system': 'blood',
    'haematopoietic system': 'blood',
    'haematopoietic': 'blood',
    'cns': 'nervous system',
    'central nervous system': 'nervous system',
    'peripheral nervous system': 'nervous system',
    'kidney': 'kidneys',
    'renal': 'kidneys',
    'hepatic': 'liver',
    'lung': 'lungs',
    'respiratory tract': 'respiratory tract',
    'respiratory system': 'respiratory system',
    'eye': 'eyes',
    'gi tract': 'gastro-intestinal tract',
    'gastrointestinal tract': 'gastro-intestinal tract',
    'gi': 'gastro-intestinal tract',
    'bone marrow': 'bone',
    'thymus': 'immune system',
    'spleen': 'immune system',
}

ORGAN_TR = {
    'blood':                    'kan',
    'nervous system':           'sinir sistemi',
    'kidneys':                  'böbrekler',
    'liver':                    'karaciğer',
    'lungs':                    'akciğerler',
    'respiratory tract':        'solunum yolu',
    'respiratory system':       'solunum sistemi',
    'eyes':                     'gözler',
    'gastro-intestinal tract':  'gastrointestinal sistem',
    'skin':                     'cilt',
    'thyroid':                  'tiroid bezi',
    'bone':                     'kemik iliği',
    'immune system':            'bağışıklık sistemi',
    'cardiovascular system':    'kardiyovasküler sistem',
    'heart':                    'kalp',
    'teeth':                    'diş',
    'upper respiratory tract':  'üst solunum yolu',
    'nervous system, eyes':     'sinir sistemi ve gözler',
}

_ORGAN_MAP = {
    'upper respiratory': 'üst solunum yolu',
    'respiratory':       'solunum sistemi',
    'nervous system':    'sinir sistemi',
    'liver':             'karaciğer',
    'kidney':            'böbrek',
    'blood':             'kan',
    'heart':             'kalp',
    'eyes':              'gözler',
    'skin':              'deri',
    'bone marrow':       'kemik iliği',
    'thyroid':           'tiroid',
}


def _normalize_organ(raw: str) -> str:
    if not raw:
        return ''
    r = raw.lower().strip()
    for en, tr in _ORGAN_MAP.items():
        if en in r:
            return tr
    return raw


def _extract_organs(h_code: str) -> List[str]:
    import re
    m = re.search(r'\(([^)]+)\)', h_code)
    if not m:
        return []
    parts = re.split(r'[,;]', m.group(1))
    return [_normalize_organ(p.strip()) for p in parts if p.strip()]


def _get_scl_cmin(comp: dict, h_code4: str) -> float | None:
    """Bileşenin SCL listesinden H kodu için c_min döndürür; yoksa None."""
    scl_raw = comp.get('scl', [])
    if isinstance(scl_raw, list):
        for s in scl_raw:
            if not isinstance(s, dict):
                continue
            sh = (s.get('h_code', '') or '').replace('*', '').strip()[:4]
            if sh == h_code4 and s.get('c_min') is not None:
                return float(s['c_min'])
    elif isinstance(scl_raw, dict):
        val = scl_raw.get(h_code4)
        if val is not None:
            return float(val)
    return None


def _scl_update(scl_organ: Dict, org: str, trig_h: str, reason: str) -> None:
    """SCL organ takibini güncelle — H372 > H373 önceliği."""
    if org not in scl_organ:
        scl_organ[org] = {'h_code': trig_h, 'reasons': [reason]}
    else:
        if trig_h == 'H372' and scl_organ[org]['h_code'] == 'H373':
            scl_organ[org]['h_code'] = 'H372'
        scl_organ[org]['reasons'].append(reason)


def _h_worse(current: 'str | None', new: str) -> str:
    """H372 > H373 önceliği."""
    return 'H372' if (current == 'H372' or new == 'H372') else new


def calculate(comps: List[Dict]) -> Dict:
    """
    STOT RE hesapla — CLP §3.9.5.4: toplamsallık yok.
    Her bileşen kendi konsantrasyonuyla tek başına Tablo 3.9.4 eşiklerine bakılır;
    farklı bileşenler toplanmaz. En ağır tekil sonuç (H372 > H373) seçilir.

    Returns:
        {
          'results': [...],
          'analytic_results': [...],
          'h_codes': [...],
          'has_general': bool,
          'general_cat1': float,  # backward compat — her zaman 0.0
          'general_cat2': float,  # backward compat — her zaman 0.0
          'warnings': []
        }
    """
    # organ → {'h_code': 'H372'|'H373'|None, 'sources': [...]}
    organ_best: Dict[str, Dict] = {}
    general_best_h: 'str | None' = None  # organ belirsiz bileşenler — en kötü tekil
    general_sources: list = []
    results: list = []
    scl_organ: Dict[str, Dict] = {}

    for c in comps:
        conc = float(c.get('concMax') or c.get('conc') or 0)
        if conc <= 0:
            continue

        for h in (c.get('hazards') or []):
            code = (h.get('h_code') or '').replace('*', '').strip()[:4]
            cat  = 1 if code == 'H372' else (2 if code == 'H373' else None)
            if cat is None:
                continue

            organs = _extract_organs(h.get('h_code') or '')
            name   = c.get('name') or c.get('cas') or ''

            scl_h372 = _get_scl_cmin(c, 'H372')
            scl_h373 = _get_scl_cmin(c, 'H373')

            if cat == 1 and (scl_h372 is not None or scl_h373 is not None):
                # SCL → bireysel değerlendirme, generic havuza katılmaz
                for org in (organs or [GENERAL_ORGAN]):
                    if scl_h372 is not None and conc >= scl_h372:
                        _scl_update(scl_organ, org, 'H372',
                                    f"{name} %{conc:.3g} ≥ SCL_H372=%{scl_h372}")
                    elif scl_h373 is not None and conc >= scl_h373:
                        _scl_update(scl_organ, org, 'H373',
                                    f"{name} %{conc:.3g} ≥ SCL_H373=%{scl_h373}")

            elif cat == 2 and scl_h373 is not None:
                for org in (organs or [GENERAL_ORGAN]):
                    if conc >= scl_h373:
                        _scl_update(scl_organ, org, 'H373',
                                    f"{name} %{conc:.3g} ≥ SCL_H373=%{scl_h373}")

            else:
                # Generic — Tablo 3.9.4, bireysel değerlendirme (toplama yok)
                this_h: 'str | None' = None
                if cat == 1:
                    if   conc >= 10.0: this_h = 'H372'
                    elif conc >=  1.0: this_h = 'H373'
                else:  # cat == 2
                    if conc >= 10.0:   this_h = 'H373'

                if this_h is None:
                    continue  # eşik altı — katkı yok

                src = {'name': name, 'conc': conc, 'cat': cat, 'h': this_h}

                if not organs:
                    general_best_h = _h_worse(general_best_h, this_h)
                    general_sources.append(src)
                else:
                    for org in organs:
                        if org not in organ_best:
                            organ_best[org] = {'h_code': None, 'sources': []}
                        organ_best[org]['h_code'] = _h_worse(organ_best[org]['h_code'], this_h)
                        organ_best[org]['sources'].append(src)

    # Organ belirsiz bileşenler: eşiği tek başına aşıyorsa muhafazakâr uygulanır
    if general_best_h is not None:
        if not organ_best:
            organ_best[GENERAL_ORGAN] = {'h_code': general_best_h, 'sources': general_sources}
        else:
            for org in organ_best:
                organ_best[org]['h_code'] = _h_worse(organ_best[org]['h_code'], general_best_h)
                organ_best[org]['sources'].extend(general_sources)

    # Generic sonuçlar — Tablo 3.9.4
    for org, data in organ_best.items():
        h = data['h_code']
        if h is None:
            continue
        srcs = data['sources']
        if h == 'H372':
            parts = [f"{s['name']} %{s['conc']:.3g} (Cat1≥%10.0→H372)"
                     for s in srcs if s['h'] == 'H372']
        else:
            parts = []
            for s in srcs:
                if s.get('cat') == 1:
                    parts.append(f"{s['name']} %{s['conc']:.3g} (Cat1 %1–%10→H373)")
                else:
                    parts.append(f"{s['name']} %{s['conc']:.3g} (Cat2≥%10.0→H373)")
        results.append({
            'h_code':  h,
            'h_class': 'STOT RE 1' if h == 'H372' else 'STOT RE 2',
            'organ':   org,
            'signal':  'Danger' if h == 'H372' else 'Warning',
            'reason':  f"{org}: {'; '.join(parts)} (CLP §3.9, Tablo 3.9.4)",
        })

    # SCL sonuçlarını ekle veya mevcut generic sonuçla birleştir
    generic_by_organ = {r['organ']: r for r in results}
    for org, sd in scl_organ.items():
        trig_h  = sd['h_code']
        scl_rsn = f"{org}: {'; '.join(sd['reasons'])} (CLP Art.10(3), Tablo 3.9.4)"
        if org not in generic_by_organ:
            results.append({
                'h_code':  trig_h,
                'h_class': 'STOT RE 1' if trig_h == 'H372' else 'STOT RE 2',
                'organ':   org,
                'signal':  'Danger' if trig_h == 'H372' else 'Warning',
                'reason':  scl_rsn,
                'scl_based': True,
            })
        elif trig_h == 'H372' and generic_by_organ[org]['h_code'] == 'H373':
            r = generic_by_organ[org]
            r['h_code']    = 'H372'
            r['h_class']   = 'STOT RE 1'
            r['signal']    = 'Danger'
            r['reason']   += f'; + SCL: {scl_rsn}'
            r['scl_based'] = True

    h_codes = list({r['h_code'] for r in results})

    return {
        'results':          results,
        'analytic_results': list(results),  # backward compat — bireysel modda aynı liste
        'h_codes':          h_codes,
        'has_general':      general_best_h is not None,
        'general_cat1':     0.0,  # toplama kaldırıldı; backward compat için 0.0
        'general_cat2':     0.0,
        'warnings':         [],
    }

File: js/engines/backup/stot_engine.js
/**
 * STOTEngine — CLP (AT) No 1272/2008 Ek I §3.9 — SDS Bölüm 11
 * STOT RE (Tekrarlı Maruziyet) hesaplama
 *
 * H372 (STOT RE 1): organ Cat1 toplamı ≥ %10.0 (generic) VEYA madde SCL_H372 ≤ konsantrasyon
 * H373 (STOT RE 2): Cat1 %1.0–%10.0 VEYA Cat2 ≥ %10.0 (generic) VEYA SCL_H373 ≤ konsantrasyon
 * SCL generic eşiğin önüne geçer — CLP Madde 10(3)
 * Kaynak: CLP Ek-1 §3.9, Tablo 3.9.4
 *
 * Events: dinler  → STOT_CALCULATE { comps }
 *         emit eder → STOT_READY { results, h_codes, ... }
 */
const STOTEngine = (() => {

  const ORGAN_MAP = {
    'upper respiratory':'üst solunum yolu', 'respiratory':'solunum sistemi',
    'nervous system':'sinir sistemi', 'liver':'karaciğer', 'kidney':'böbrek',
    'blood':'kan', 'heart':'kalp', 'eyes':'gözler', 'skin':'deri',
    'bone marrow':'kemik iliği', 'thyroid':'tiroid',
  };

  function normalizeOrgan(raw) {
    if (!raw) return null;
    const r = raw.toLowerCase().trim();
    for (const [en, tr] of Object.entries(ORGAN_MAP)) {
      if (r.includes(en)) return tr;
    }
    return raw;
  }

  function extractOrgans(hcode) {
    const m = hcode.match(/\(([^)]+)\)/);
    if (!m) return [];
    return m[1].split(/[,;]/).map(s => normalizeOrgan(s.trim())).filter(Boolean);
  }

  /** Bileşenin SCL listesinden H kodu için c_min döndürür; yoksa null. */
  function getSCLCmin(comp, hCode4) {
    const scl = comp.scl;
    if (!scl) return null;
    if (Array.isArray(scl)) {
      for (const s of scl) {
        if (!s || typeof s !== 'object') continue;
        const sh = (s.h_code || '').replace(/\*/g, '').trim().substring(0, 4);
        if (sh === hCode4 && s.c_min != null) return parseFloat(s.c_min);
      }
      return null;
    }
    if (typeof scl === 'object') {
      const val = scl[hCode4];
      return val != null ? parseFloat(val) : null;
    }
    return null;
  }

  /** SCL organ takibini güncelle — H372 > H373 önceliği. */
  function sclUpdate(sclOrgan, org, trigH, reason) {
    if (!sclOrgan[org]) {
      sclOrgan[org] = { h: trigH, reasons: [reason] };
    } else {
      if (trigH === 'H372' && sclOrgan[org].h === 'H373') sclOrgan[org].h = 'H372';
      sclOrgan[org].reasons.push(reason);
    }
  }

  function calculate(comps) {
    const organSums = {};
    let generalCat1 = 0, generalCat2 = 0;
    const results = [], analyticResults = [];
    const sclOrgan = {};  // org → { h: 'H372'|'H373', reasons: [] }

    for (const c of comps) {
      const conc = parseFloat(c.concMax || c.conc) || 0;
      if (conc <= 0) continue;

      const name = c.name || c.cas || '';

      for (const h of (c.hazards || [])) {
        const code = (h.h_code || '').replace(/[*\s]/g,'').substring(0,4);
        const cat = code === 'H372' ? 1 : code === 'H373' ? 2 : null;
        if (!cat) continue;

        const organs = extractOrgans(h.h_code || '');

        // SCL kontrolü — CLP Art. 10(3): SCL generic eşiğin yerini alır
        const sclH372 = getSCLCmin(c, 'H372');
        const sclH373 = getSCLCmin(c, 'H373');

        if (cat === 1 && (sclH372 !== null || sclH373 !== null)) {
          // SCL tanımlı → bireysel değerlendirme, generic havuza katılmaz
          const targets = organs.length > 0 ? organs : ['Genel (organ belirsiz)'];
          for (const org of targets) {
            if (sclH372 !== null && conc >= sclH372) {
              sclUpdate(sclOrgan, org, 'H372',
                `${name} %${+conc.toFixed(3)} ≥ SCL_H372=%${sclH372}`);
            } else if (sclH373 !== null && conc >= sclH373) {
              sclUpdate(sclOrgan, org, 'H373',
                `${name} %${+conc.toFixed(3)} ≥ SCL_H373=%${sclH373}`);
            }
            // SCL eşiği altındaysa katkı yok
          }

        } else if (cat === 2 && sclH373 !== null) {
          // H373 + SCL → bireysel değerlendirme
          const targets = organs.length > 0 ? organs : ['Genel (organ belirsiz)'];
          for (const org of targets) {
            if (conc >= sclH373) {
              sclUpdate(sclOrgan, org, 'H373',
                `${name} %${+conc.toFixed(3)} ≥ SCL_H373=%${sclH373}`);
            }
          }

        } else {
          // Generic additive havuz — mevcut davranış
          if (organs.length === 0) {
            if (cat === 1) generalCat1 += conc;
            else            generalCat2 += conc;
          } else {
            organs.forEach(org => {
              if (!organSums[org]) organSums[org] = { cat1:0, cat2:0, sources:[] };
              if (cat === 1) organSums[org].cat1 += conc;
              else           organSums[org].cat2 += conc;
              organSums[org].sources.push({ name, conc, cat });
            });
          }
        }
      }
    }

    // Organ belirsiz maddeler tüm organlara muhafazakâr olarak eklenir
    const allOrgans = Object.keys(organSums);
    for (const org of allOrgans) {
      organSums[org].cat1 += generalCat1;
      organSums[org].cat2 += generalCat2;
    }

    // Generic sonuçlar — Tablo 3.9.4
    for (const [org, sums] of Object.entries(organSums)) {
      if (sums.cat1 >= 10.0) {
        results.push({ h:'H372', h_class:'STOT RE 1', organ: org, signal:'Danger',
          reason:`${org}: STOT RE 1 toplamı %${sums.cat1.toFixed(1)} ≥ %10.0 (KKDİK Ek-2, Tablo 3.9.4)` });
      } else if (sums.cat1 >= 1.0 || sums.cat2 >= 10.0) {
        const parts = [];
        if (sums.cat1 >= 1.0) parts.push(`STOT RE 1 toplamı %${sums.cat1.toFixed(1)} (%1.0–%10.0 → H373)`);
        if (sums.cat2 >= 10.0) parts.push(`STOT RE 2 toplamı %${sums.cat2.toFixed(1)} ≥ %10.0`);
        results.push({ h:'H373', h_class:'STOT RE 2', organ: org, signal:'Warning',
          reason:`${org}: ${parts.join('; ')} (KKDİK Ek-2, Tablo 3.9.4)` });
      }
    }

    // SCL sonuçlarını ekle veya mevcut generic sonuçla birleştir
    const genericByOrgan = {};
    for (const r of results) genericByOrgan[r.organ] = r;

    for (const [org, sd] of Object.entries(sclOrgan)) {
      const trigH  = sd.h;
      const sclRsn = `${org}: ${sd.reasons.join('; ')} (CLP Art.10(3), Tablo 3.9.4)`;
      if (!genericByOrgan[org]) {
        results.push({
          h: trigH, h_class: trigH === 'H372' ? 'STOT RE 1' : 'STOT RE 2',
          organ: org, signal: trigH === 'H372' ? 'Danger' : 'Warning',
          reason: sclRsn, scl_based: true,
        });
      } else if (trigH === 'H372' && genericByOrgan[org].h === 'H373') {
        // SCL H372 generic H373'ü geçersiz kılar
        const r = genericByOrgan[org];
        r.h       = 'H372';
        r.h_class = 'STOT RE 1';
        r.signal  = 'Danger';
        r.reason += `; + SCL: ${sclRsn}`;
        r.scl_based = true;
      }
    }

    // Analitik mod: organ belirsiz hariç — generic
    for (const [org, sums] of Object.entries(organSums)) {
      const c1 = sums.cat1 - generalCat1;
      const c2 = sums.cat2 - generalCat2;
      if (c1 >= 10.0) {
        analyticResults.push({ h:'H372', h_class:'STOT RE 1', organ: org, signal:'Danger',
          reason:`${org}: Eşleşen Cat1=%${c1.toFixed(1)} ≥ %10.0`, general_excl: generalCat1 });
      } else if (c1 >= 1.0 || c2 >= 10.0) {
        const parts = [];
        if (c1 >= 1.0) parts.push(`Eşleşen Cat1=%${c1.toFixed(1)} (%1.0–%10.0 → H373)`);
        if (c2 >= 10.0) parts.push(`Eşleşen Cat2=%${c2.toFixed(1)} ≥ %10.0`);
        analyticResults.push({ h:'H373', h_class:'STOT RE 2', organ: org, signal:'Warning',
          reason:`${org}: ${parts.join('; ')}`, general_excl: generalCat1 });
      }
    }

    // SCL sonuçları analitik listede de yer alır
    for (const [org, sd] of Object.entries(sclOrgan)) {
      const trigH = sd.h;
      analyticResults.push({
        h: trigH, h_class: trigH === 'H372' ? 'STOT RE 1' : 'STOT RE 2',
        organ: org, signal: trigH === 'H372' ? 'Danger' : 'Warning',
        reason: `${org}: ${sd.reasons.join('; ')} (SCL)`,
        scl_based: true,
      });
    }

    // Organ belirsiz — genel havuz (hiç organ eşleşmesi yoksa)
    if (Object.keys(organSums).length === 0) {
      if (generalCat1 >= 10.0) {
        results.push({ h:'H372', h_class:'STOT RE 1', organ:'Genel (organ belirsiz)', signal:'Danger',
          reason:`Genel: Cat1=%${generalCat1.toFixed(1)} ≥ %10.0 (Tablo 3.9.4)` });
      } else if (generalCat1 >= 1.0 || generalCat2 >= 10.0) {
        const parts = [];
        if (generalCat1 >= 1.0) parts.push(`Cat1=%${generalCat1.toFixed(1)} (%1.0–%10.0 → H373)`);
        if (generalCat2 >= 10.0) parts.push(`Cat2=%${generalCat2.toFixed(1)} ≥ %10.0`);
        results.push({ h:'H373', h_class:'STOT RE 2', organ:'Genel (organ belirsiz)', signal:'Warning',
          reason:`Genel: ${parts.join('; ')} (Tablo 3.9.4)` });
      }
    }

    const h_codes = [...new Set(results.map(r => r.h))];

    return {
      results, analytic_results: analyticResults, h_codes,
      has_general: generalCat1 > 0 || generalCat2 > 0,
      general_cat1: generalCat1, general_cat2: generalCat2,
      warnings: [],
    };
  }

  function init() {
    if (typeof EventBus !== 'undefined') {
      EventBus.on('STOT_CALCULATE', ({ comps }) => {
        const result = calculate(comps);
        EventBus.emit('STOT_READY', result);
      });
    }
    console.log('[STOTEngine] init OK');
  }

  return { init, calculate };
})();
